File: app/src/main/java/meugeninua/appfromscratch/ui/activities/base/fragments/base/binding/BaseBinding.java
package meugeninua.appfromscratch.ui.activities.base.fragments.base.binding;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.util.SparseArrayCompat;
import android.view.View;

import java.lang.ref.WeakReference;

public abstract class BaseBinding {

    private WeakReference<View> rootViewRef;
    private SparseArrayCompat<WeakReference<View>> childrenViewRefs;

    public void attachView(final View view) {
        rootViewRef = new WeakReference<>(view);
        childrenViewRefs = new SparseArrayCompat<>();
    }

    public void detachView() {
        rootViewRef = null;
        childrenViewRefs.clear();
        childrenViewRefs = null;
    }

    @Nullable
    private <V extends View> V getNullable(final int id) {
        View rootView = rootViewRef == null ? null : rootViewRef.get();
        if (rootView == null) {
            return null;
        }
        WeakReference<View> childViewRef = childrenViewRefs.get(id);
        View view = childViewRef == null ? null : childViewRef.get();
        if (view == null) {
            view = rootView.findViewById(id);
            if (view != null) {
                childrenViewRefs.put(id, new WeakReference<View>(view));
            }
        }
        return (V) view;
    }

    @NonNull
    public <V extends View> V get(final int id) {
        final V view = getNullable(id);
        if (view == null) {
            throw new IllegalArgumentException("View not found");
        }
        return view;
    }

    public boolean has(final int id) {
        final View view = getNullable(id);
        return view != null;
    }
}

File: app/src/main/java/meugeninua/appfromscratch/ui/activities/base/fragments/base/binding/BaseBinding.kt
package meugeninua.appfromscratch.ui.activities.base.fragments.base.binding

import android.support.v4.util.SparseArrayCompat
import android.view.View

import java.lang.ref.WeakReference

abstract class BaseBinding {

    private var rootViewRef: WeakReference<View>? = null
    private var childrenViewRefs: SparseArrayCompat<WeakReference<View>>? = null

    fun attachView(view: View) {
        rootViewRef = WeakReference(view)
        childrenViewRefs = SparseArrayCompat()
    }

    fun detachView() {
        rootViewRef = null
        childrenViewRefs?.clear()
        childrenViewRefs = null
    }

    private fun <V : View> getNullable(id: Int): V? {
        val rootViewRef = this.rootViewRef ?: return null
        val childrenViewRefs = this.childrenViewRefs ?: return null
        val rootView = rootViewRef.get() ?: return null
        val childViewRef = childrenViewRefs.get(id)
        var view: View? = childViewRef?.get()
        if (view == null) {
            view = rootView.findViewById(id)
            if (view != null) {
                childrenViewRefs.put(id, WeakReference(view))
            }
        }
        return view as V?
    }

    operator fun <V : View> get(id: Int): V {
        return getNullable(id) ?: throw IllegalArgumentException("View not found")
    }

    fun has(id: Int): Boolean {
        val view = getNullable<View>(id)
        return view != null
    }
}
